### PR TITLE
Filter bucket list by region

### LINKED ISSUE #12: Filter bucket list by region
I have found this tool super useful, thank you for making it! Due to the problem described I first thought that the program did not work at all, as I could not open any of my buckets. This is why I think it's worth fixing. 

The list of buckets show all buckets in the aws account for all regions. This makes it seem like they can all be accessed. However when trying to open a bucket that resides in a different region than specified (`ap-northeast-1` by default), it fails with the error `ERROR: Failed to load objects`.  It would be really useful to show a `invalid bucket region` or something in this case to avoid confusion. 

If you want I can make a pull request on this :) I have zero experience with rust so it would be interesting to give it a shot.

### PR DESCRIPTION
**Pull request by henrikskog**
Fixes #12 

If you want I can get the locations in parallell. I'm a complete novice in Rust so there are probably ways of writing this neater as well!

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -68,10 +68,41 @@ impl Client {
...
+                            other => other,
+                        };
+
+                        if constraint_str == self.region {
+                            buckets_in_region.push(bucket.clone());
+                        }
+                    }
+                }
+                Err(e) => {
+                    tracing::error!("Failed to get bucket location: {:?}", e);
```
Comment (henrikskog) [inline src/client.rs:94]: Should this be handled maybe?

Comment (lusingander) [inline src/client.rs:94]: It would seem fine to return an AppError for failed requests, just like with other requests.

File: src/client.rs [hunk 2]
```diff
@@ -68,10 +68,41 @@ impl Client {
...
+            match bucket_location {
+                Ok(location) => {
+                    if let Some(constraint) = location.location_constraint() {
+                        let constraint_str = match constraint.as_str() {
+                            "" => "us-east-1", // https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/get_bucket_location/struct.GetBucketLocationOutput.html#method.location_constraint
+                            other => other,
+                        };
+
+                        if constraint_str == self.region {
+                            buckets_in_region.push(bucket.clone());
```
Comment (lusingander) [inline src/client.rs:89]: Probably don't need to clone here.

File: src/client.rs [hunk 3]
```diff
@@ -68,10 +68,41 @@ impl Client {
...
+                .send()
+                .await;
+
+            match bucket_location {
+                Ok(location) => {
+                    if let Some(constraint) = location.location_constraint() {
+                        let constraint_str = match constraint.as_str() {
+                            "" => "us-east-1", // https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/get_bucket_location/struct.GetBucketLocationOutput.html#method.location_constraint
+                            other => other,
+                        };
```
Comment (lusingander) [inline src/client.rs:86]: Could you please comment that `location.location_constraint()` returns `Some("")` and not `None`? (The comments in the aws sdk seem vague...)

**Review by lusingander — CHANGES_REQUESTED** [review]
Thanks for fixing! I’ve left a few minor comments.

> If you want I can get the locations in parallell.

I agree that reducing the number of requests corresponding to the number of buckets is an improvement we should consider. However, I don't think it needs to be addressed in this pull request.

In practice, bucket names are unique, and regions rarely change, so I think it's acceptable to cache this information.

**Comment by henrikskog** [comment]
Thank you for the review! I'll hopefully take a look at it tonight or tomorrow :)

**Comment by henrikskog** [comment]
> In practice, bucket names are unique, and regions rarely change, so I think it's acceptable to cache this information.

You want me to take a shot at implementing this?

File: src/client.rs [hunk 4]
```diff
@@ -81,17 +85,22 @@ impl Client {
...
+                            buckets_in_region.push(bucket);
                         }
                     }
                 }
                 Err(e) => {
-                    tracing::error!("Failed to get bucket location: {:?}", e);
+                    return Err(AppError::new(
+                        format!("Failed to get bucket location for bucket {}", bucket.name),
+                        e,
+                    ));
```
Comment (lusingander) [inline src/client.rs]: it seems to be better to use `.map_err` and `?`.

```rs
let bucket_location = self.client.get_bucket_location().bucket(bucket.name.as_str()).send().await.;
let location = bucket_location.map_err(|e| AppError::new("...", e))?;
```

**Review by lusingander — CHANGES_REQUESTED** [review]

**Comment by lusingander** [comment]
> > In practice, bucket names are unique, and regions rarely change, so I think it's acceptable to cache this information.
> 
> You want me to take a shot at implementing this?

You don't necessarily have to address this issue. But if you'd like to take it on, you are welcome to do so.

File: src/client.rs [hunk 5]
```diff
@@ -81,38 +82,35 @@ impl Client {
...
-                }
-                Err(e) => {
-                    return Err(AppError::new(
-                        format!("Failed to get bucket location for bucket {}", bucket.name),
-                        e,
-                    ));
+            let location = bucket_location.map_err(|e| {
+                AppError::new(
+                    format!("Failed to get bucket location for bucket {}", bucket.name),
+                    e,
```
Comment (henrikskog) [inline src/client.rs:88]: I assume this will end up in `.stu/error.log` as well as showing the message to the user?

**Comment by henrikskog** [comment]
> > > In practice, bucket names are unique, and regions rarely change, so I think it's acceptable to cache this information.
> > 
> > 
> > You want me to take a shot at implementing this?
> 
> You don't necessarily have to address this issue. But if you'd like to take it on, you are welcome to do so.

I'd like to 👍

File: src/client.rs (hunk 5 above)
Comment (lusingander) [reply src/client.rs:88]: Yes, it will be logged when we notify with `AppError`.

**Review by lusingander — APPROVED** [review]

**Comment by lusingander** [comment]
Thank you!

**Pull request merged**